Alternate formats for extracting data from MDX query

DJHig opened the pull request:
Hello!
First, thank you for the really useful TM1py project!
I wanted to pull data in a different format than what your normal execute_mdx gives so I added a couple of additional formats.  I am driving multiple synched Dygraphs from a single query with these new methods. I added a get_data method that works like execute_mdx, but could also be used as a template for a similar get_data_from_view, allowing data to be returned in any format. The get_data could be used as a drop-in replacement for execute_mdx since the additional arguments are optional and default is how execute_mdx works. (If you want to do this, I can change the order of the arguments so calls with positional arguments will still work).

Let me know what you think. I can give you some sample code so you can see the different formats in action. I have one that queries the default "Planning Sample" server from TM1 so it should work for everyone.
Thanks!

MariusWirtz commented on TM1py/Services/CellService.py:
```diff
@@ -179,6 +179,111 @@ def execute_mdx(self, mdx, cell_properties=None, top=None):
         finally:
             self.delete_cellset(cellset_id=cellset_id)
 
+    def get_data(self, mdx, format=None, cell_properites=None, dim_properties=None, top=None):
```
typo in cell_properties

MariusWirtz replied:
Maybe dim_properties should be called elem_properties?

MariusWirtz replied:
Maybe rename format parameter. Unnecessary use of builtin names should be avoided

MariusWirtz commented on TM1py/Services/CellService.py:
```diff
@@ -179,6 +179,111 @@ def execute_mdx(self, mdx, cell_properties=None, top=None):
+    def get_data(self, mdx, format=None, cell_properites=None, dim_properties=None, top=None):
+        """ Execute an MDX query and return the results in the format specified
+
+        :param mdx_or_view: A string that is an MDX Query or view name
```
wrong parameter in docstring

MariusWirtz commented on TM1py/Utils/Utils.py:
```diff
@@ -111,6 +111,122 @@ def build_content_from_cellset(raw_cellset_as_dict, cell_properties, top=None):
+    dimensionality = len(raw_cellset_as_dict['Axes'])
+    cardinality = [raw_cellset_as_dict['Axes'][axis]['Cardinality'] for axis in range(dimensionality)]
+
+    headerMap = build_headers_from_cellset(raw_cellset_as_dict)
```
header_map

MariusWirtz commented on TM1py/Utils/Utils.py:
```diff
@@ -111,6 +111,122 @@ def build_content_from_cellset(raw_cellset_as_dict, cell_properties, top=None):
+            row = [xHeader]
+            for y in range(cardinality[1]):
+                cell_addr = (x + cardinality[0] * y + cardinality[0] * cardinality[1] * z)
+                row.append(float("{0:.1f}".format(raw_cellset_as_dict['Cells'][cell_addr]['Value'] or 0)))
```
Would be nice if precision was configurable I think

MariusWirtz commented on TM1py/Utils/Utils.py:
```diff
@@ -111,6 +111,122 @@ def build_content_from_cellset(raw_cellset_as_dict, cell_properties, top=None):
 
 
+def build_arrays_from_cellset(raw_cellset_as_dict):
+    """ Transform raw 1,2 or 3-dimension cellset data into concise dictionary
```
Maybe it makes sense to add a comment what we can use this response type for

MariusWirtz commented on TM1py/Services/CellService.py:
```diff
@@ -179,6 +179,111 @@ def execute_mdx(self, mdx, cell_properties=None, top=None):
+            memory_file = StringIO(raw_csv)
+            return pd.read_csv(memory_file, sep=',')
+        else:
+            if not cell_properties:
```
I think lines 218 to 221 are obsolete

DJHig replied:
Oh, I only added these here since build_content_from_cellset needs the cell_properties to know which properties to add. It could be derived from the data and then it does not need them. Also, not sure if top is needed there since the query would be limiting the data.

DJHig replied:
These lines are only needed to make sure the Utils.build_content_from_cellset method has the 'Ordinal' passed in the cell_properties (actually, to make sure it matches what the query in extract_cellset_full used.) I think Utils.build_content_from_cellset could derive that from the data and it does not need to be passed. I'm not sure it needs the top argument either, since the query already limited these values. What do you think about changing that function to take only raw_cellset_as_dict and value_precision, like the two new ones?

DJHig replied:
Sorry, didn't mean to send duplicate comment.

MariusWirtz replied:
You are absolutely right. There is no need to query the Ordinal property from the cell. This needs to be fixed. 

Yes. It doesn’t necessarily need the top argument. We added it, so that theoretically the utility function is independent from the REST call. 
I think it makes sense, to add the value-precision as an argument. 

I wouldn’t want to remove the cell_properties argument. Sometimes it's useful to query other properties from the cell, like RuleDerived or CellAnnotation.

MariusWirtz commented on TM1py/Utils/Utils.py:
```diff
@@ -111,6 +111,122 @@ def build_content_from_cellset(raw_cellset_as_dict, cell_properties, top=None):
+    :param raw_cellset_as_dict: raw data from TM1
+    :return: dict : { titles: [], headers: [axis][], cells: { Page0: [  [column name, column values], [], ... ], ...} }
+    """
+    dimensionality = len(raw_cellset_as_dict['Axes'])
```
Maybe outsource these two lines into seperate function to avoid code duplication:
dimensionality = len(raw_cellset_as_dict['Axes'])
cardinality = [raw_cellset_as_dict['Axes'][axis]['Cardinality'] for axis in range(dimensionality)]

Or maybe build_headers_from_cellset could also return dimensionality and cardinality?

DJHig commented:
Hi Marius,
Thank you for the review and comments. I made the requested changes.
I added an example 'cells' output from both of my added output formats. The arrays one was for an earlier charting library I was evaluating that needed the rows to be arrays, rather than the columns (like dygraphs or pandas). The arrays format is also useful for grids that usually want the data by rows.

I did not add value_precision to the build_content_from_cellset, but I can add that if you would like. 

I don't the get_data to be a single function and it is fine with me to split it into the smaller ones by specific format. I was mainly trying to reduce the code duplication with having to repeat them for Views, in addition to MDX queries. I am currently only using MDX queries so I do not need the Views now anyway.

Regarding dygraphs, I can try to come up with an example. Yes, I was using Flask, but ran into a weird bug so I switched to Sanic.

Thanks!

DJHig commented:
I also updated the sample in TM1py-samples to match the new argument names.
